contracts/BodhiCandy.sol

```diff
@@ -0,0 +1,65 @@
+pragma solidity ^0.4.18;
+
+import './libs/SafeMath.sol';
+
+contract BodhiCandy {
+    using SafeMath for uint256;
+
+    // Amount in Satoshi to deposit
+    uint256 public constant depositAmount = 1000000;
+
+    // Amount of blocks it takes for the last person to deposit to win if no others deposit within the length.
+    uint256 public constant winningBlockLength = 30;
+
+    address public lastDepositer;
+    uint256 public lastDepositBlock;
+    uint256 public currentBalance;
+
+    event UserWon(address indexed winner, uint256 amountWon);
+    event Returned(uint256 returned);
+
+    function deposit() external payable {
+        require(msg.value >= depositAmount);
+
+        // Calculate any amount over the depositAmount to return to user
+        uint256 excessDeposit = 0;
+        if (msg.value > depositAmount) {
+            excessDeposit = msg.value.sub(depositAmount);
+        }
+
+        // Last depositer wins
+        if (lastDepositer != address(0) 
+            && lastDepositBlock != 0 
+            && block.number.sub(lastDepositBlock) >= winningBlockLength) {
+
+            uint256 amountWon = currentBalance.add(excessDeposit);
```

Comment:
I know you want to prevent to send payment twice (one for remainnnig, one for reward), but you code below has bug because you combine the two together.

Reply:
This is to prevent a reentrancy attack. Can you explain the bug?

example:
currentBalance = .05
30 blocks pass
another user deposits() .06 qtum
excessDeposit = .06 - .01 = .05
total qtum in contract = .11
amountWon = .05 (curr bal) + .05 = .1
transfer(amountWon = .1)
remaining qtum in contract = .11 - .1 = .01
set currentBalance = depositAmount = .01